<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="twitter:card" content="summary_large_image" /><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>(in progress) Purines, Pyrimidines, and Convolutional Neural Networks | Simon Says</title>
<meta name="generator" content="Jekyll v4.1.1" />
<meta property="og:title" content="(in progress) Purines, Pyrimidines, and Convolutional Neural Networks" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A sketch of CNNs for homotypic motif density localization (finding clusters of the same type of sequence motif)." />
<meta property="og:description" content="A sketch of CNNs for homotypic motif density localization (finding clusters of the same type of sequence motif)." />
<link rel="canonical" href="https://simonlevine.github.io/simonsays/jupyter/2021/02/17/DNA_CNN.html" />
<meta property="og:url" content="https://simonlevine.github.io/simonsays/jupyter/2021/02/17/DNA_CNN.html" />
<meta property="og:site_name" content="Simon Says" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2021-02-17T00:00:00-06:00" />
<script type="application/ld+json">
{"url":"https://simonlevine.github.io/simonsays/jupyter/2021/02/17/DNA_CNN.html","@type":"BlogPosting","headline":"(in progress) Purines, Pyrimidines, and Convolutional Neural Networks","dateModified":"2021-02-17T00:00:00-06:00","datePublished":"2021-02-17T00:00:00-06:00","mainEntityOfPage":{"@type":"WebPage","@id":"https://simonlevine.github.io/simonsays/jupyter/2021/02/17/DNA_CNN.html"},"description":"A sketch of CNNs for homotypic motif density localization (finding clusters of the same type of sequence motif).","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/simonsays/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="https://simonlevine.github.io/simonsays/feed.xml" title="Simon Says" /><link rel="shortcut icon" type="image/x-icon" href="/simonsays/images/favicon.ico"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/Primer/15.2.0/primer.css" integrity="sha512-xTz2ys4coGAOz8vuV1NcQBkgVmKhsSEtjbqyMJbBHRplFuvKIUo6xhLHpAyPt9mfR6twHJgn9OgVLuqOvjeBhg==" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha512-1PKOgIY59xJ8Co8+NE6FZ+LOAZKjy+KY8iq0G4B3CyeY6wYHN3yt9PW0XpSriVlkMXe40PTKnXrLnZ9+fkDaog==" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/katex.min.css" integrity="sha512-h7nl+xz8wgDlNM4NqKEM4F1NkIRS17M9+uJwIGwuo8vGqIl4BhuCKdxjWEINm+xyrUjNCnK5dCrhM0sj+wTIXw==" crossorigin="anonymous" />
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/katex.min.js" integrity="sha512-/CMIhXiDA3m2c9kzRyd97MTb3MC6OVnx4TElQ7fkkoRghwDf6gi41gaT1PwF270W6+J60uTmwgeRpNpJdRV6sg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/contrib/auto-render.min.js" integrity="sha512-Do7uJAaHZm5OLrIv/yN4w0iG1dbu01kzdMNnFfu/mAqgUk6Nniv2JYHcwH+cNwjqgLcqcuBBk+JRvprLVI8azg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A==" crossorigin="anonymous"></script>
    <script>
    document.addEventListener("DOMContentLoaded", function() {
        renderMathInElement( document.body, {
        delimiters: [
            {left: "$$", right: "$$", display: true},
            {left: "[%", right: "%]", display: true},
            {left: "$", right: "$", display: false}
        ]}
        );
    });
    </script>


<script>
function wrap_img(fn) {
    if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading") {
        var elements = document.querySelectorAll(".post img");
        Array.prototype.forEach.call(elements, function(el, i) {
            if (el.getAttribute("title") && (el.className != "emoji")) {
                const caption = document.createElement('figcaption');
                var node = document.createTextNode(el.getAttribute("title"));
                caption.appendChild(node);
                const wrapper = document.createElement('figure');
                wrapper.className = 'image';
                el.parentNode.insertBefore(wrapper, el);
                el.parentNode.removeChild(el);
                wrapper.appendChild(el);
                wrapper.appendChild(caption);
            }
        });
    } else { document.addEventListener('DOMContentLoaded', fn); }
}
window.onload = wrap_img;
</script>

<script>
    document.addEventListener("DOMContentLoaded", function(){
    // add link icon to anchor tags
    var elem = document.querySelectorAll(".anchor-link")
    elem.forEach(e => (e.innerHTML = '<i class="fas fa-link fa-xs"></i>'));
    });
</script>
</head>
<body><header class="site-header">

  <div class="wrapper"><a class="site-title" rel="author" href="/simonsays/">Simon Says</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/simonsays/about/">About Me</a><a class="page-link" href="/simonsays/search/">Search</a><a class="page-link" href="/simonsays/categories/">Tags</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">(in progress) Purines, Pyrimidines, and Convolutional Neural Networks</h1><p class="page-description">A sketch of CNNs for homotypic motif density localization (finding clusters of the same type of sequence motif).</p><p class="post-meta post-meta-title"><time class="dt-published" datetime="2021-02-17T00:00:00-06:00" itemprop="datePublished">
        Feb 17, 2021
      </time>
       â€¢ <span class="read-time" title="Estimated read time">
    
    
      10 min read
    
</span></p>

    
      <p class="category-tags"><i class="fas fa-tags category-tags-icon"></i></i> 
      
        <a class="category-tags-link" href="/simonsays/categories/#jupyter">jupyter</a>
        
      
      </p>
    

    
      
        <div class="pb-5 d-flex flex-wrap flex-justify-end">
          <div class="px-2">

    <a href="https://github.com/simonlevine/simonsays/tree/master/_notebooks/2021-02-17-DNA_CNN.ipynb" role="button" target="_blank">
<img class="notebook-badge-image" src="/simonsays/assets/badges/github.svg" alt="View On GitHub">
    </a>
</div>

          <div class="px-2">
    <a href="https://mybinder.org/v2/gh/simonlevine/simonsays/master?filepath=_notebooks%2F2021-02-17-DNA_CNN.ipynb" target="_blank">
        <img class="notebook-badge-image" src="/simonsays/assets/badges/binder.svg" alt="Open In Binder"/>
    </a>
</div>

          <div class="px-2">
    <a href="https://colab.research.google.com/github/simonlevine/simonsays/blob/master/_notebooks/2021-02-17-DNA_CNN.ipynb" target="_blank">
        <img class="notebook-badge-image" src="/simonsays/assets/badges/colab.svg" alt="Open In Colab"/>
    </a>
</div>
        </div>
      </header>

  <div class="post-content e-content" itemprop="articleBody">
    <ul class="section-nav">
<li class="toc-entry toc-h1"><a href="#Background">Background </a>
<ul>
<li class="toc-entry toc-h2"><a href="#Biology">Biology </a></li>
<li class="toc-entry toc-h2"><a href="#Dataset">Dataset </a></li>
<li class="toc-entry toc-h2"><a href="#Machine-Learning">Machine Learning </a>
<ul>
<li class="toc-entry toc-h3"><a href="#Task:">Task: </a></li>
</ul>
</li>
</ul>
</li>
</ul><!--
#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: _notebooks/2021-02-17-DNA_CNN.ipynb
-->

<div class="container" id="notebook-container">
        
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Background">
<a class="anchor" href="#Background" aria-hidden="true"><span class="octicon octicon-link"></span></a>Background<a class="anchor-link" href="#Background"> </a>
</h1>
<p><em>This notebook is adapted from the CS-185, Spring 2021 course at Brown University</em> (<a href="http://cs.brown.edu/courses/csci1850">http://cs.brown.edu/courses/csci1850</a>)</p>
<h2 id="Biology">
<a class="anchor" href="#Biology" aria-hidden="true"><span class="octicon octicon-link"></span></a>Biology<a class="anchor-link" href="#Biology"> </a>
</h2>
<p>Transcription Factors (TFs) are an important class of protein that bind to sections of DNA. This binding process is mediated by their propensity to attach to particular sequence motifs, or patterns in the DNA sequence. Often, we humans refer to binding as important regulatory regions, as associated transcription factors play a large role in upregulating and/or downregulating sequence transcription. In other words, finding the location of such motifs on a given sequence allows us to glean a functional understanding from structure alone. If an algorithm can be trained to detect relevant motifs given a cluster with high fidelity, then the algorithm could obviate hand-tuned feature identification, and even predict likely TF binding sites from unknown sequences. In fact, this is the exact problem we will solve in this exercise.</p>
<h2 id="Dataset">
<a class="anchor" href="#Dataset" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dataset<a class="anchor-link" href="#Dataset"> </a>
</h2>
<p>The small dataset provided actually is <em>simulated</em> homotypic motif density localization data. For more about how to generate such data, see the Kundaje Lab's DragoNN (<a href="http://kundajelab.github.io/dragonn/">http://kundajelab.github.io/dragonn/</a>), and if you want, try to beat their results with, perhaps, the GAN from my previous post ðŸ˜‰. The nice thing about this dataset is that it is already formatted for easy loading. Each "sequence" is actually a $1500-mer$, and, given their <em>in silico</em> provenance, we can assume ðŸŒŸ perfect coverage ðŸŒŸ and ðŸŒ¸ no sequencing errors ðŸŒ¸ for the purpose of this exercise. We can take a peek at the training split below.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span> 
</pre></div>

    </div>
</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">'./data_dna_cnn/hw1_data.npz'</span><span class="p">)</span>
<span class="n">train_seq</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">'train_seq'</span><span class="p">]</span>
<span class="n">train_y</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">'train_y'</span><span class="p">]</span>
<span class="n">train_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">train_seq</span><span class="p">,</span><span class="n">train_y</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],[</span><span class="s1">'X'</span><span class="p">,</span><span class="s1">'y'</span><span class="p">])))</span>
<span class="n">train_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>X</th>
      <th>y</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TGTTACGCTATCGCGAACAGTATGTTCAAGTGGGCGTTCGCTAGGA...</td>
      <td>[True]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CAAACTTCTTCAAAGGATCCCAGATGCGGTTAAATCATACCACCAA...</td>
      <td>[True]</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GACCATGCCTTGAGGCGTATACATTTGTTGTAGTTGAATTACTACG...</td>
      <td>[True]</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AGTATTACGATCATAGAGTATCAGAGTTGTTTTGTAACAGCGTCGG...</td>
      <td>[False]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CTTCAGATACGTAATTACTCTATAATTTACAAACAACCACGTATCC...</td>
      <td>[False]</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AGGAGTTTCTATTTCAGAAGTGAACGGTGACAAAAAGTCCACTCAA...</td>
      <td>[True]</td>
    </tr>
    <tr>
      <th>6</th>
      <td>GGGTCGAAAAGTATGACGATCCAATCATTAGAAGTGAGGTTCACCT...</td>
      <td>[False]</td>
    </tr>
    <tr>
      <th>7</th>
      <td>AATATTTAATTTATCTCCGACCAGTCAAAGATACTGTCTTGAACTG...</td>
      <td>[True]</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GGCGCCTTTTTTCTCAAAAGAAAGTGATGTTGTTTTGCATTTCTAT...</td>
      <td>[True]</td>
    </tr>
    <tr>
      <th>9</th>
      <td>TCGCCGGATGGACTGCTACCTTGTTGGACTAATGGTATATTTTTCC...</td>
      <td>[True]</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">train_df</span><span class="o">.</span><span class="n">X</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>array([1500])</pre>
</div>

</div>

</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">'dark_background'</span><span class="p">)</span>
<span class="n">train_df</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">'hist'</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">train_df</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&lt;matplotlib.axes._subplots.AxesSubplot at 0x7fe49424e310&gt;</pre>
</div>

</div>

<div class="output_area">


<div class="output_svg output_subarea ">
<?xml version="1.0" encoding="utf-8" standalone="no"?>
&lt;!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
  "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"&gt;
<!-- Created with matplotlib (https://matplotlib.org/) -->
<svg height="290.511pt" version="1.1" viewbox="0 0 395.328125 290.511" width="395.328125pt" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
 <defs>
  <style type="text/css">
*{stroke-linecap:butt;stroke-linejoin:round;}
  </style>
 </defs>
 <g id="figure_1">
  <g id="patch_1">
   <path d="M 0 290.511 
L 395.328125 290.511 
L 395.328125 0 
L 0 0 
z
"></path>
  </g>
  <g id="axes_1">
   <g id="patch_2">
    <path d="M 53.328125 266.632875 
L 388.128125 266.632875 
L 388.128125 49.192875 
L 53.328125 49.192875 
z
"></path>
   </g>
   <g id="patch_3">
    <path clip-path="url(#p13d87deb40)" d="M 68.546307 266.632875 
L 98.98267 266.632875 
L 98.98267 61.607715 
L 68.546307 61.607715 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_4">
    <path clip-path="url(#p13d87deb40)" d="M 98.98267 266.632875 
L 129.419034 266.632875 
L 129.419034 266.632875 
L 98.98267 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_5">
    <path clip-path="url(#p13d87deb40)" d="M 129.419034 266.632875 
L 159.855398 266.632875 
L 159.855398 266.632875 
L 129.419034 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_6">
    <path clip-path="url(#p13d87deb40)" d="M 159.855398 266.632875 
L 190.291761 266.632875 
L 190.291761 266.632875 
L 159.855398 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_7">
    <path clip-path="url(#p13d87deb40)" d="M 190.291761 266.632875 
L 220.728125 266.632875 
L 220.728125 266.632875 
L 190.291761 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_8">
    <path clip-path="url(#p13d87deb40)" d="M 220.728125 266.632875 
L 251.164489 266.632875 
L 251.164489 266.632875 
L 220.728125 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_9">
    <path clip-path="url(#p13d87deb40)" d="M 251.164489 266.632875 
L 281.600852 266.632875 
L 281.600852 266.632875 
L 251.164489 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_10">
    <path clip-path="url(#p13d87deb40)" d="M 281.600852 266.632875 
L 312.037216 266.632875 
L 312.037216 266.632875 
L 281.600852 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_11">
    <path clip-path="url(#p13d87deb40)" d="M 312.037216 266.632875 
L 342.47358 266.632875 
L 342.47358 266.632875 
L 312.037216 266.632875 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="patch_12">
    <path clip-path="url(#p13d87deb40)" d="M 342.47358 266.632875 
L 372.909943 266.632875 
L 372.909943 59.547161 
L 342.47358 59.547161 
z
" style="fill:#8dd3c7;"></path>
   </g>
   <g id="matplotlib.axis_1">
    <g id="xtick_1">
     <g id="line2d_1">
      <defs>
       <path d="M 0 0 
L 0 3.5 
" id="m812d83f0db" style="stroke:#ffffff;stroke-width:0.8;"></path>
      </defs>
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="68.546307" xlink:href="#m812d83f0db" y="266.632875"></use>
      </g>
     </g>
     <g id="text_1">
      <!-- 0.0 -->
      <defs>
       <path d="M 31.78125 66.40625 
Q 24.171875 66.40625 20.328125 58.90625 
Q 16.5 51.421875 16.5 36.375 
Q 16.5 21.390625 20.328125 13.890625 
Q 24.171875 6.390625 31.78125 6.390625 
Q 39.453125 6.390625 43.28125 13.890625 
Q 47.125 21.390625 47.125 36.375 
Q 47.125 51.421875 43.28125 58.90625 
Q 39.453125 66.40625 31.78125 66.40625 
z
M 31.78125 74.21875 
Q 44.046875 74.21875 50.515625 64.515625 
Q 56.984375 54.828125 56.984375 36.375 
Q 56.984375 17.96875 50.515625 8.265625 
Q 44.046875 -1.421875 31.78125 -1.421875 
Q 19.53125 -1.421875 13.0625 8.265625 
Q 6.59375 17.96875 6.59375 36.375 
Q 6.59375 54.828125 13.0625 64.515625 
Q 19.53125 74.21875 31.78125 74.21875 
z
" id="DejaVuSans-48"></path>
       <path d="M 10.6875 12.40625 
L 21 12.40625 
L 21 0 
L 10.6875 0 
z
" id="DejaVuSans-46"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(60.594744 281.231312)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-48"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-46"></use>
       <use x="95.410156" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="xtick_2">
     <g id="line2d_2">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="129.419034" xlink:href="#m812d83f0db" y="266.632875"></use>
      </g>
     </g>
     <g id="text_2">
      <!-- 0.2 -->
      <defs>
       <path d="M 19.1875 8.296875 
L 53.609375 8.296875 
L 53.609375 0 
L 7.328125 0 
L 7.328125 8.296875 
Q 12.9375 14.109375 22.625 23.890625 
Q 32.328125 33.6875 34.8125 36.53125 
Q 39.546875 41.84375 41.421875 45.53125 
Q 43.3125 49.21875 43.3125 52.78125 
Q 43.3125 58.59375 39.234375 62.25 
Q 35.15625 65.921875 28.609375 65.921875 
Q 23.96875 65.921875 18.8125 64.3125 
Q 13.671875 62.703125 7.8125 59.421875 
L 7.8125 69.390625 
Q 13.765625 71.78125 18.9375 73 
Q 24.125 74.21875 28.421875 74.21875 
Q 39.75 74.21875 46.484375 68.546875 
Q 53.21875 62.890625 53.21875 53.421875 
Q 53.21875 48.921875 51.53125 44.890625 
Q 49.859375 40.875 45.40625 35.40625 
Q 44.1875 33.984375 37.640625 27.21875 
Q 31.109375 20.453125 19.1875 8.296875 
z
" id="DejaVuSans-50"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(121.467472 281.231312)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-48"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-46"></use>
       <use x="95.410156" xlink:href="#DejaVuSans-50"></use>
      </g>
     </g>
    </g>
    <g id="xtick_3">
     <g id="line2d_3">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="190.291761" xlink:href="#m812d83f0db" y="266.632875"></use>
      </g>
     </g>
     <g id="text_3">
      <!-- 0.4 -->
      <defs>
       <path d="M 37.796875 64.3125 
L 12.890625 25.390625 
L 37.796875 25.390625 
z
M 35.203125 72.90625 
L 47.609375 72.90625 
L 47.609375 25.390625 
L 58.015625 25.390625 
L 58.015625 17.1875 
L 47.609375 17.1875 
L 47.609375 0 
L 37.796875 0 
L 37.796875 17.1875 
L 4.890625 17.1875 
L 4.890625 26.703125 
z
" id="DejaVuSans-52"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(182.340199 281.231312)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-48"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-46"></use>
       <use x="95.410156" xlink:href="#DejaVuSans-52"></use>
      </g>
     </g>
    </g>
    <g id="xtick_4">
     <g id="line2d_4">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="251.164489" xlink:href="#m812d83f0db" y="266.632875"></use>
      </g>
     </g>
     <g id="text_4">
      <!-- 0.6 -->
      <defs>
       <path d="M 33.015625 40.375 
Q 26.375 40.375 22.484375 35.828125 
Q 18.609375 31.296875 18.609375 23.390625 
Q 18.609375 15.53125 22.484375 10.953125 
Q 26.375 6.390625 33.015625 6.390625 
Q 39.65625 6.390625 43.53125 10.953125 
Q 47.40625 15.53125 47.40625 23.390625 
Q 47.40625 31.296875 43.53125 35.828125 
Q 39.65625 40.375 33.015625 40.375 
z
M 52.59375 71.296875 
L 52.59375 62.3125 
Q 48.875 64.0625 45.09375 64.984375 
Q 41.3125 65.921875 37.59375 65.921875 
Q 27.828125 65.921875 22.671875 59.328125 
Q 17.53125 52.734375 16.796875 39.40625 
Q 19.671875 43.65625 24.015625 45.921875 
Q 28.375 48.1875 33.59375 48.1875 
Q 44.578125 48.1875 50.953125 41.515625 
Q 57.328125 34.859375 57.328125 23.390625 
Q 57.328125 12.15625 50.6875 5.359375 
Q 44.046875 -1.421875 33.015625 -1.421875 
Q 20.359375 -1.421875 13.671875 8.265625 
Q 6.984375 17.96875 6.984375 36.375 
Q 6.984375 53.65625 15.1875 63.9375 
Q 23.390625 74.21875 37.203125 74.21875 
Q 40.921875 74.21875 44.703125 73.484375 
Q 48.484375 72.75 52.59375 71.296875 
z
" id="DejaVuSans-54"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(243.212926 281.231312)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-48"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-46"></use>
       <use x="95.410156" xlink:href="#DejaVuSans-54"></use>
      </g>
     </g>
    </g>
    <g id="xtick_5">
     <g id="line2d_5">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="312.037216" xlink:href="#m812d83f0db" y="266.632875"></use>
      </g>
     </g>
     <g id="text_5">
      <!-- 0.8 -->
      <defs>
       <path d="M 31.78125 34.625 
Q 24.75 34.625 20.71875 30.859375 
Q 16.703125 27.09375 16.703125 20.515625 
Q 16.703125 13.921875 20.71875 10.15625 
Q 24.75 6.390625 31.78125 6.390625 
Q 38.8125 6.390625 42.859375 10.171875 
Q 46.921875 13.96875 46.921875 20.515625 
Q 46.921875 27.09375 42.890625 30.859375 
Q 38.875 34.625 31.78125 34.625 
z
M 21.921875 38.8125 
Q 15.578125 40.375 12.03125 44.71875 
Q 8.5 49.078125 8.5 55.328125 
Q 8.5 64.0625 14.71875 69.140625 
Q 20.953125 74.21875 31.78125 74.21875 
Q 42.671875 74.21875 48.875 69.140625 
Q 55.078125 64.0625 55.078125 55.328125 
Q 55.078125 49.078125 51.53125 44.71875 
Q 48 40.375 41.703125 38.8125 
Q 48.828125 37.15625 52.796875 32.3125 
Q 56.78125 27.484375 56.78125 20.515625 
Q 56.78125 9.90625 50.3125 4.234375 
Q 43.84375 -1.421875 31.78125 -1.421875 
Q 19.734375 -1.421875 13.25 4.234375 
Q 6.78125 9.90625 6.78125 20.515625 
Q 6.78125 27.484375 10.78125 32.3125 
Q 14.796875 37.15625 21.921875 38.8125 
z
M 18.3125 54.390625 
Q 18.3125 48.734375 21.84375 45.5625 
Q 25.390625 42.390625 31.78125 42.390625 
Q 38.140625 42.390625 41.71875 45.5625 
Q 45.3125 48.734375 45.3125 54.390625 
Q 45.3125 60.0625 41.71875 63.234375 
Q 38.140625 66.40625 31.78125 66.40625 
Q 25.390625 66.40625 21.84375 63.234375 
Q 18.3125 60.0625 18.3125 54.390625 
z
" id="DejaVuSans-56"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(304.085653 281.231312)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-48"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-46"></use>
       <use x="95.410156" xlink:href="#DejaVuSans-56"></use>
      </g>
     </g>
    </g>
    <g id="xtick_6">
     <g id="line2d_6">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="372.909943" xlink:href="#m812d83f0db" y="266.632875"></use>
      </g>
     </g>
     <g id="text_6">
      <!-- 1.0 -->
      <defs>
       <path d="M 12.40625 8.296875 
L 28.515625 8.296875 
L 28.515625 63.921875 
L 10.984375 60.40625 
L 10.984375 69.390625 
L 28.421875 72.90625 
L 38.28125 72.90625 
L 38.28125 8.296875 
L 54.390625 8.296875 
L 54.390625 0 
L 12.40625 0 
z
" id="DejaVuSans-49"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(364.958381 281.231312)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-49"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-46"></use>
       <use x="95.410156" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
   </g>
   <g id="matplotlib.axis_2">
    <g id="ytick_1">
     <g id="line2d_7">
      <defs>
       <path d="M 0 0 
L -3.5 0 
" id="m4b52eadb9f" style="stroke:#ffffff;stroke-width:0.8;"></path>
      </defs>
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="266.632875"></use>
      </g>
     </g>
     <g id="text_7">
      <!-- 0 -->
      <g style="fill:#ffffff;" transform="translate(39.965625 270.432094)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_2">
     <g id="line2d_8">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="240.875945"></use>
      </g>
     </g>
     <g id="text_8">
      <!-- 250 -->
      <defs>
       <path d="M 10.796875 72.90625 
L 49.515625 72.90625 
L 49.515625 64.59375 
L 19.828125 64.59375 
L 19.828125 46.734375 
Q 21.96875 47.46875 24.109375 47.828125 
Q 26.265625 48.1875 28.421875 48.1875 
Q 40.625 48.1875 47.75 41.5 
Q 54.890625 34.8125 54.890625 23.390625 
Q 54.890625 11.625 47.5625 5.09375 
Q 40.234375 -1.421875 26.90625 -1.421875 
Q 22.3125 -1.421875 17.546875 -0.640625 
Q 12.796875 0.140625 7.71875 1.703125 
L 7.71875 11.625 
Q 12.109375 9.234375 16.796875 8.0625 
Q 21.484375 6.890625 26.703125 6.890625 
Q 35.15625 6.890625 40.078125 11.328125 
Q 45.015625 15.765625 45.015625 23.390625 
Q 45.015625 31 40.078125 35.4375 
Q 35.15625 39.890625 26.703125 39.890625 
Q 22.75 39.890625 18.8125 39.015625 
Q 14.890625 38.140625 10.796875 36.28125 
z
" id="DejaVuSans-53"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(27.240625 244.675164)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-50"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-53"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_3">
     <g id="line2d_9">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="215.119016"></use>
      </g>
     </g>
     <g id="text_9">
      <!-- 500 -->
      <g style="fill:#ffffff;" transform="translate(27.240625 218.918234)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-53"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-48"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_4">
     <g id="line2d_10">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="189.362086"></use>
      </g>
     </g>
     <g id="text_10">
      <!-- 750 -->
      <defs>
       <path d="M 8.203125 72.90625 
L 55.078125 72.90625 
L 55.078125 68.703125 
L 28.609375 0 
L 18.3125 0 
L 43.21875 64.59375 
L 8.203125 64.59375 
z
" id="DejaVuSans-55"></path>
      </defs>
      <g style="fill:#ffffff;" transform="translate(27.240625 193.161305)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-55"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-53"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_5">
     <g id="line2d_11">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="163.605156"></use>
      </g>
     </g>
     <g id="text_11">
      <!-- 1000 -->
      <g style="fill:#ffffff;" transform="translate(20.878125 167.404375)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-49"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-48"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-48"></use>
       <use x="190.869141" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_6">
     <g id="line2d_12">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="137.848227"></use>
      </g>
     </g>
     <g id="text_12">
      <!-- 1250 -->
      <g style="fill:#ffffff;" transform="translate(20.878125 141.647446)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-49"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-50"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-53"></use>
       <use x="190.869141" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_7">
     <g id="line2d_13">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="112.091297"></use>
      </g>
     </g>
     <g id="text_13">
      <!-- 1500 -->
      <g style="fill:#ffffff;" transform="translate(20.878125 115.890516)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-49"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-53"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-48"></use>
       <use x="190.869141" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_8">
     <g id="line2d_14">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="86.334368"></use>
      </g>
     </g>
     <g id="text_14">
      <!-- 1750 -->
      <g style="fill:#ffffff;" transform="translate(20.878125 90.133586)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-49"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-55"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-53"></use>
       <use x="190.869141" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="ytick_9">
     <g id="line2d_15">
      <g>
       <use style="fill:#ffffff;stroke:#ffffff;stroke-width:0.8;" x="53.328125" xlink:href="#m4b52eadb9f" y="60.577438"></use>
      </g>
     </g>
     <g id="text_15">
      <!-- 2000 -->
      <g style="fill:#ffffff;" transform="translate(20.878125 64.376657)scale(0.1 -0.1)">
       <use xlink:href="#DejaVuSans-50"></use>
       <use x="63.623047" xlink:href="#DejaVuSans-48"></use>
       <use x="127.246094" xlink:href="#DejaVuSans-48"></use>
       <use x="190.869141" xlink:href="#DejaVuSans-48"></use>
      </g>
     </g>
    </g>
    <g id="text_16">
     <!-- Frequency -->
     <defs>
      <path d="M 9.8125 72.90625 
L 51.703125 72.90625 
L 51.703125 64.59375 
L 19.671875 64.59375 
L 19.671875 43.109375 
L 48.578125 43.109375 
L 48.578125 34.8125 
L 19.671875 34.8125 
L 19.671875 0 
L 9.8125 0 
z
" id="DejaVuSans-70"></path>
      <path d="M 41.109375 46.296875 
Q 39.59375 47.171875 37.8125 47.578125 
Q 36.03125 48 33.890625 48 
Q 26.265625 48 22.1875 43.046875 
Q 18.109375 38.09375 18.109375 28.8125 
L 18.109375 0 
L 9.078125 0 
L 9.078125 54.6875 
L 18.109375 54.6875 
L 18.109375 46.1875 
Q 20.953125 51.171875 25.484375 53.578125 
Q 30.03125 56 36.53125 56 
Q 37.453125 56 38.578125 55.875 
Q 39.703125 55.765625 41.0625 55.515625 
z
" id="DejaVuSans-114"></path>
      <path d="M 56.203125 29.59375 
L 56.203125 25.203125 
L 14.890625 25.203125 
Q 15.484375 15.921875 20.484375 11.0625 
Q 25.484375 6.203125 34.421875 6.203125 
Q 39.59375 6.203125 44.453125 7.46875 
Q 49.3125 8.734375 54.109375 11.28125 
L 54.109375 2.78125 
Q 49.265625 0.734375 44.1875 -0.34375 
Q 39.109375 -1.421875 33.890625 -1.421875 
Q 20.796875 -1.421875 13.15625 6.1875 
Q 5.515625 13.8125 5.515625 26.8125 
Q 5.515625 40.234375 12.765625 48.109375 
Q 20.015625 56 32.328125 56 
Q 43.359375 56 49.78125 48.890625 
Q 56.203125 41.796875 56.203125 29.59375 
z
M 47.21875 32.234375 
Q 47.125 39.59375 43.09375 43.984375 
Q 39.0625 48.390625 32.421875 48.390625 
Q 24.90625 48.390625 20.390625 44.140625 
Q 15.875 39.890625 15.1875 32.171875 
z
" id="DejaVuSans-101"></path>
      <path d="M 14.796875 27.296875 
Q 14.796875 17.390625 18.875 11.75 
Q 22.953125 6.109375 30.078125 6.109375 
Q 37.203125 6.109375 41.296875 11.75 
Q 45.40625 17.390625 45.40625 27.296875 
Q 45.40625 37.203125 41.296875 42.84375 
Q 37.203125 48.484375 30.078125 48.484375 
Q 22.953125 48.484375 18.875 42.84375 
Q 14.796875 37.203125 14.796875 27.296875 
z
M 45.40625 8.203125 
Q 42.578125 3.328125 38.25 0.953125 
Q 33.9375 -1.421875 27.875 -1.421875 
Q 17.96875 -1.421875 11.734375 6.484375 
Q 5.515625 14.40625 5.515625 27.296875 
Q 5.515625 40.1875 11.734375 48.09375 
Q 17.96875 56 27.875 56 
Q 33.9375 56 38.25 53.625 
Q 42.578125 51.265625 45.40625 46.390625 
L 45.40625 54.6875 
L 54.390625 54.6875 
L 54.390625 -20.796875 
L 45.40625 -20.796875 
z
" id="DejaVuSans-113"></path>
      <path d="M 8.5 21.578125 
L 8.5 54.6875 
L 17.484375 54.6875 
L 17.484375 21.921875 
Q 17.484375 14.15625 20.5 10.265625 
Q 23.53125 6.390625 29.59375 6.390625 
Q 36.859375 6.390625 41.078125 11.03125 
Q 45.3125 15.671875 45.3125 23.6875 
L 45.3125 54.6875 
L 54.296875 54.6875 
L 54.296875 0 
L 45.3125 0 
L 45.3125 8.40625 
Q 42.046875 3.421875 37.71875 1 
Q 33.40625 -1.421875 27.6875 -1.421875 
Q 18.265625 -1.421875 13.375 4.4375 
Q 8.5 10.296875 8.5 21.578125 
z
M 31.109375 56 
z
" id="DejaVuSans-117"></path>
      <path d="M 54.890625 33.015625 
L 54.890625 0 
L 45.90625 0 
L 45.90625 32.71875 
Q 45.90625 40.484375 42.875 44.328125 
Q 39.84375 48.1875 33.796875 48.1875 
Q 26.515625 48.1875 22.3125 43.546875 
Q 18.109375 38.921875 18.109375 30.90625 
L 18.109375 0 
L 9.078125 0 
L 9.078125 54.6875 
L 18.109375 54.6875 
L 18.109375 46.1875 
Q 21.34375 51.125 25.703125 53.5625 
Q 30.078125 56 35.796875 56 
Q 45.21875 56 50.046875 50.171875 
Q 54.890625 44.34375 54.890625 33.015625 
z
" id="DejaVuSans-110"></path>
      <path d="M 48.78125 52.59375 
L 48.78125 44.1875 
Q 44.96875 46.296875 41.140625 47.34375 
Q 37.3125 48.390625 33.40625 48.390625 
Q 24.65625 48.390625 19.8125 42.84375 
Q 14.984375 37.3125 14.984375 27.296875 
Q 14.984375 17.28125 19.8125 11.734375 
Q 24.65625 6.203125 33.40625 6.203125 
Q 37.3125 6.203125 41.140625 7.25 
Q 44.96875 8.296875 48.78125 10.40625 
L 48.78125 2.09375 
Q 45.015625 0.34375 40.984375 -0.53125 
Q 36.96875 -1.421875 32.421875 -1.421875 
Q 20.0625 -1.421875 12.78125 6.34375 
Q 5.515625 14.109375 5.515625 27.296875 
Q 5.515625 40.671875 12.859375 48.328125 
Q 20.21875 56 33.015625 56 
Q 37.15625 56 41.109375 55.140625 
Q 45.0625 54.296875 48.78125 52.59375 
z
" id="DejaVuSans-99"></path>
      <path d="M 32.171875 -5.078125 
Q 28.375 -14.84375 24.75 -17.8125 
Q 21.140625 -20.796875 15.09375 -20.796875 
L 7.90625 -20.796875 
L 7.90625 -13.28125 
L 13.1875 -13.28125 
Q 16.890625 -13.28125 18.9375 -11.515625 
Q 21 -9.765625 23.484375 -3.21875 
L 25.09375 0.875 
L 2.984375 54.6875 
L 12.5 54.6875 
L 29.59375 11.921875 
L 46.6875 54.6875 
L 56.203125 54.6875 
z
" id="DejaVuSans-121"></path>
     </defs>
     <g style="fill:#ffffff;" transform="translate(14.798438 184.210531)rotate(-90)scale(0.1 -0.1)">
      <use xlink:href="#DejaVuSans-70"></use>
      <use x="57.410156" xlink:href="#DejaVuSans-114"></use>
      <use x="98.492188" xlink:href="#DejaVuSans-101"></use>
      <use x="160.015625" xlink:href="#DejaVuSans-113"></use>
      <use x="223.492188" xlink:href="#DejaVuSans-117"></use>
      <use x="286.871094" xlink:href="#DejaVuSans-101"></use>
      <use x="348.394531" xlink:href="#DejaVuSans-110"></use>
      <use x="411.773438" xlink:href="#DejaVuSans-99"></use>
      <use x="466.753906" xlink:href="#DejaVuSans-121"></use>
     </g>
    </g>
   </g>
   <g id="patch_13">
    <path d="M 53.328125 266.632875 
L 53.328125 49.192875 
" style="fill:none;stroke:#ffffff;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;"></path>
   </g>
   <g id="patch_14">
    <path d="M 388.128125 266.632875 
L 388.128125 49.192875 
" style="fill:none;stroke:#ffffff;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;"></path>
   </g>
   <g id="patch_15">
    <path d="M 53.328125 266.632875 
L 388.128125 266.632875 
" style="fill:none;stroke:#ffffff;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;"></path>
   </g>
   <g id="patch_16">
    <path d="M 53.328125 49.192875 
L 388.128125 49.192875 
" style="fill:none;stroke:#ffffff;stroke-linecap:square;stroke-linejoin:miter;stroke-width:0.8;"></path>
   </g>
   <g id="text_17">
    <!-- True     2010 -->
    <defs>
     <path d="M -0.296875 72.90625 
L 61.375 72.90625 
L 61.375 64.59375 
L 35.5 64.59375 
L 35.5 0 
L 25.59375 0 
L 25.59375 64.59375 
L -0.296875 64.59375 
z
" id="DejaVuSans-84"></path>
     <path id="DejaVuSans-32"></path>
    </defs>
    <g style="fill:#ffffff;" transform="translate(182.31125 16.318125)scale(0.12 -0.12)">
     <use xlink:href="#DejaVuSans-84"></use>
     <use x="60.865234" xlink:href="#DejaVuSans-114"></use>
     <use x="101.978516" xlink:href="#DejaVuSans-117"></use>
     <use x="165.357422" xlink:href="#DejaVuSans-101"></use>
     <use x="226.880859" xlink:href="#DejaVuSans-32"></use>
     <use x="258.667969" xlink:href="#DejaVuSans-32"></use>
     <use x="290.455078" xlink:href="#DejaVuSans-32"></use>
     <use x="322.242188" xlink:href="#DejaVuSans-32"></use>
     <use x="354.029297" xlink:href="#DejaVuSans-32"></use>
     <use x="385.816406" xlink:href="#DejaVuSans-50"></use>
     <use x="449.439453" xlink:href="#DejaVuSans-48"></use>
     <use x="513.0625" xlink:href="#DejaVuSans-49"></use>
     <use x="576.685547" xlink:href="#DejaVuSans-48"></use>
    </g>
    <!-- False    1990 -->
    <defs>
     <path d="M 34.28125 27.484375 
Q 23.390625 27.484375 19.1875 25 
Q 14.984375 22.515625 14.984375 16.5 
Q 14.984375 11.71875 18.140625 8.90625 
Q 21.296875 6.109375 26.703125 6.109375 
Q 34.1875 6.109375 38.703125 11.40625 
Q 43.21875 16.703125 43.21875 25.484375 
L 43.21875 27.484375 
z
M 52.203125 31.203125 
L 52.203125 0 
L 43.21875 0 
L 43.21875 8.296875 
Q 40.140625 3.328125 35.546875 0.953125 
Q 30.953125 -1.421875 24.3125 -1.421875 
Q 15.921875 -1.421875 10.953125 3.296875 
Q 6 8.015625 6 15.921875 
Q 6 25.140625 12.171875 29.828125 
Q 18.359375 34.515625 30.609375 34.515625 
L 43.21875 34.515625 
L 43.21875 35.40625 
Q 43.21875 41.609375 39.140625 45 
Q 35.0625 48.390625 27.6875 48.390625 
Q 23 48.390625 18.546875 47.265625 
Q 14.109375 46.140625 10.015625 43.890625 
L 10.015625 52.203125 
Q 14.9375 54.109375 19.578125 55.046875 
Q 24.21875 56 28.609375 56 
Q 40.484375 56 46.34375 49.84375 
Q 52.203125 43.703125 52.203125 31.203125 
z
" id="DejaVuSans-97"></path>
     <path d="M 9.421875 75.984375 
L 18.40625 75.984375 
L 18.40625 0 
L 9.421875 0 
z
" id="DejaVuSans-108"></path>
     <path d="M 44.28125 53.078125 
L 44.28125 44.578125 
Q 40.484375 46.53125 36.375 47.5 
Q 32.28125 48.484375 27.875 48.484375 
Q 21.1875 48.484375 17.84375 46.4375 
Q 14.5 44.390625 14.5 40.28125 
Q 14.5 37.15625 16.890625 35.375 
Q 19.28125 33.59375 26.515625 31.984375 
L 29.59375 31.296875 
Q 39.15625 29.25 43.1875 25.515625 
Q 47.21875 21.78125 47.21875 15.09375 
Q 47.21875 7.46875 41.1875 3.015625 
Q 35.15625 -1.421875 24.609375 -1.421875 
Q 20.21875 -1.421875 15.453125 -0.5625 
Q 10.6875 0.296875 5.421875 2 
L 5.421875 11.28125 
Q 10.40625 8.6875 15.234375 7.390625 
Q 20.0625 6.109375 24.8125 6.109375 
Q 31.15625 6.109375 34.5625 8.28125 
Q 37.984375 10.453125 37.984375 14.40625 
Q 37.984375 18.0625 35.515625 20.015625 
Q 33.0625 21.96875 24.703125 23.78125 
L 21.578125 24.515625 
Q 13.234375 26.265625 9.515625 29.90625 
Q 5.8125 33.546875 5.8125 39.890625 
Q 5.8125 47.609375 11.28125 51.796875 
Q 16.75 56 26.8125 56 
Q 31.78125 56 36.171875 55.265625 
Q 40.578125 54.546875 44.28125 53.078125 
z
" id="DejaVuSans-115"></path>
     <path d="M 10.984375 1.515625 
L 10.984375 10.5 
Q 14.703125 8.734375 18.5 7.8125 
Q 22.3125 6.890625 25.984375 6.890625 
Q 35.75 6.890625 40.890625 13.453125 
Q 46.046875 20.015625 46.78125 33.40625 
Q 43.953125 29.203125 39.59375 26.953125 
Q 35.25 24.703125 29.984375 24.703125 
Q 19.046875 24.703125 12.671875 31.3125 
Q 6.296875 37.9375 6.296875 49.421875 
Q 6.296875 60.640625 12.9375 67.421875 
Q 19.578125 74.21875 30.609375 74.21875 
Q 43.265625 74.21875 49.921875 64.515625 
Q 56.59375 54.828125 56.59375 36.375 
Q 56.59375 19.140625 48.40625 8.859375 
Q 40.234375 -1.421875 26.421875 -1.421875 
Q 22.703125 -1.421875 18.890625 -0.6875 
Q 15.09375 0.046875 10.984375 1.515625 
z
M 30.609375 32.421875 
Q 37.25 32.421875 41.125 36.953125 
Q 45.015625 41.5 45.015625 49.421875 
Q 45.015625 57.28125 41.125 61.84375 
Q 37.25 66.40625 30.609375 66.40625 
Q 23.96875 66.40625 20.09375 61.84375 
Q 16.21875 57.28125 16.21875 49.421875 
Q 16.21875 41.5 20.09375 36.953125 
Q 23.96875 32.421875 30.609375 32.421875 
z
" id="DejaVuSans-57"></path>
    </defs>
    <g style="fill:#ffffff;" transform="translate(182.226875 29.7555)scale(0.12 -0.12)">
     <use xlink:href="#DejaVuSans-70"></use>
     <use x="57.378906" xlink:href="#DejaVuSans-97"></use>
     <use x="118.658203" xlink:href="#DejaVuSans-108"></use>
     <use x="146.441406" xlink:href="#DejaVuSans-115"></use>
     <use x="198.541016" xlink:href="#DejaVuSans-101"></use>
     <use x="260.064453" xlink:href="#DejaVuSans-32"></use>
     <use x="291.851562" xlink:href="#DejaVuSans-32"></use>
     <use x="323.638672" xlink:href="#DejaVuSans-32"></use>
     <use x="355.425781" xlink:href="#DejaVuSans-32"></use>
     <use x="387.212891" xlink:href="#DejaVuSans-49"></use>
     <use x="450.835938" xlink:href="#DejaVuSans-57"></use>
     <use x="514.458984" xlink:href="#DejaVuSans-57"></use>
     <use x="578.082031" xlink:href="#DejaVuSans-48"></use>
    </g>
    <!-- Name: y, dtype: int64 -->
    <defs>
     <path d="M 9.8125 72.90625 
L 23.09375 72.90625 
L 55.421875 11.921875 
L 55.421875 72.90625 
L 64.984375 72.90625 
L 64.984375 0 
L 51.703125 0 
L 19.390625 60.984375 
L 19.390625 0 
L 9.8125 0 
z
" id="DejaVuSans-78"></path>
     <path d="M 52 44.1875 
Q 55.375 50.25 60.0625 53.125 
Q 64.75 56 71.09375 56 
Q 79.640625 56 84.28125 50.015625 
Q 88.921875 44.046875 88.921875 33.015625 
L 88.921875 0 
L 79.890625 0 
L 79.890625 32.71875 
Q 79.890625 40.578125 77.09375 44.375 
Q 74.3125 48.1875 68.609375 48.1875 
Q 61.625 48.1875 57.5625 43.546875 
Q 53.515625 38.921875 53.515625 30.90625 
L 53.515625 0 
L 44.484375 0 
L 44.484375 32.71875 
Q 44.484375 40.625 41.703125 44.40625 
Q 38.921875 48.1875 33.109375 48.1875 
Q 26.21875 48.1875 22.15625 43.53125 
Q 18.109375 38.875 18.109375 30.90625 
L 18.109375 0 
L 9.078125 0 
L 9.078125 54.6875 
L 18.109375 54.6875 
L 18.109375 46.1875 
Q 21.1875 51.21875 25.484375 53.609375 
Q 29.78125 56 35.6875 56 
Q 41.65625 56 45.828125 52.96875 
Q 50 49.953125 52 44.1875 
z
" id="DejaVuSans-109"></path>
     <path d="M 11.71875 12.40625 
L 22.015625 12.40625 
L 22.015625 0 
L 11.71875 0 
z
M 11.71875 51.703125 
L 22.015625 51.703125 
L 22.015625 39.3125 
L 11.71875 39.3125 
z
" id="DejaVuSans-58"></path>
     <path d="M 11.71875 12.40625 
L 22.015625 12.40625 
L 22.015625 4 
L 14.015625 -11.625 
L 7.71875 -11.625 
L 11.71875 4 
z
" id="DejaVuSans-44"></path>
     <path d="M 45.40625 46.390625 
L 45.40625 75.984375 
L 54.390625 75.984375 
L 54.390625 0 
L 45.40625 0 
L 45.40625 8.203125 
Q 42.578125 3.328125 38.25 0.953125 
Q 33.9375 -1.421875 27.875 -1.421875 
Q 17.96875 -1.421875 11.734375 6.484375 
Q 5.515625 14.40625 5.515625 27.296875 
Q 5.515625 40.1875 11.734375 48.09375 
Q 17.96875 56 27.875 56 
Q 33.9375 56 38.25 53.625 
Q 42.578125 51.265625 45.40625 46.390625 
z
M 14.796875 27.296875 
Q 14.796875 17.390625 18.875 11.75 
Q 22.953125 6.109375 30.078125 6.109375 
Q 37.203125 6.109375 41.296875 11.75 
Q 45.40625 17.390625 45.40625 27.296875 
Q 45.40625 37.203125 41.296875 42.84375 
Q 37.203125 48.484375 30.078125 48.484375 
Q 22.953125 48.484375 18.875 42.84375 
Q 14.796875 37.203125 14.796875 27.296875 
z
" id="DejaVuSans-100"></path>
     <path d="M 18.3125 70.21875 
L 18.3125 54.6875 
L 36.8125 54.6875 
L 36.8125 47.703125 
L 18.3125 47.703125 
L 18.3125 18.015625 
Q 18.3125 11.328125 20.140625 9.421875 
Q 21.96875 7.515625 27.59375 7.515625 
L 36.8125 7.515625 
L 36.8125 0 
L 27.59375 0 
Q 17.1875 0 13.234375 3.875 
Q 9.28125 7.765625 9.28125 18.015625 
L 9.28125 47.703125 
L 2.6875 47.703125 
L 2.6875 54.6875 
L 9.28125 54.6875 
L 9.28125 70.21875 
z
" id="DejaVuSans-116"></path>
     <path d="M 18.109375 8.203125 
L 18.109375 -20.796875 
L 9.078125 -20.796875 
L 9.078125 54.6875 
L 18.109375 54.6875 
L 18.109375 46.390625 
Q 20.953125 51.265625 25.265625 53.625 
Q 29.59375 56 35.59375 56 
Q 45.5625 56 51.78125 48.09375 
Q 58.015625 40.1875 58.015625 27.296875 
Q 58.015625 14.40625 51.78125 6.484375 
Q 45.5625 -1.421875 35.59375 -1.421875 
Q 29.59375 -1.421875 25.265625 0.953125 
Q 20.953125 3.328125 18.109375 8.203125 
z
M 48.6875 27.296875 
Q 48.6875 37.203125 44.609375 42.84375 
Q 40.53125 48.484375 33.40625 48.484375 
Q 26.265625 48.484375 22.1875 42.84375 
Q 18.109375 37.203125 18.109375 27.296875 
Q 18.109375 17.390625 22.1875 11.75 
Q 26.265625 6.109375 33.40625 6.109375 
Q 40.53125 6.109375 44.609375 11.75 
Q 48.6875 17.390625 48.6875 27.296875 
z
" id="DejaVuSans-112"></path>
     <path d="M 9.421875 54.6875 
L 18.40625 54.6875 
L 18.40625 0 
L 9.421875 0 
z
M 9.421875 75.984375 
L 18.40625 75.984375 
L 18.40625 64.59375 
L 9.421875 64.59375 
z
" id="DejaVuSans-105"></path>
    </defs>
    <g style="fill:#ffffff;" transform="translate(155.135 43.192875)scale(0.12 -0.12)">
     <use xlink:href="#DejaVuSans-78"></use>
     <use x="74.804688" xlink:href="#DejaVuSans-97"></use>
     <use x="136.083984" xlink:href="#DejaVuSans-109"></use>
     <use x="233.496094" xlink:href="#DejaVuSans-101"></use>
     <use x="295.019531" xlink:href="#DejaVuSans-58"></use>
     <use x="328.710938" xlink:href="#DejaVuSans-32"></use>
     <use x="360.498047" xlink:href="#DejaVuSans-121"></use>
     <use x="419.677734" xlink:href="#DejaVuSans-44"></use>
     <use x="451.464844" xlink:href="#DejaVuSans-32"></use>
     <use x="483.251953" xlink:href="#DejaVuSans-100"></use>
     <use x="546.728516" xlink:href="#DejaVuSans-116"></use>
     <use x="585.9375" xlink:href="#DejaVuSans-121"></use>
     <use x="645.117188" xlink:href="#DejaVuSans-112"></use>
     <use x="708.59375" xlink:href="#DejaVuSans-101"></use>
     <use x="770.117188" xlink:href="#DejaVuSans-58"></use>
     <use x="803.808594" xlink:href="#DejaVuSans-32"></use>
     <use x="835.595703" xlink:href="#DejaVuSans-105"></use>
     <use x="863.378906" xlink:href="#DejaVuSans-110"></use>
     <use x="926.757812" xlink:href="#DejaVuSans-116"></use>
     <use x="965.966797" xlink:href="#DejaVuSans-54"></use>
     <use x="1029.589844" xlink:href="#DejaVuSans-52"></use>
    </g>
   </g>
  </g>
 </g>
 <defs>
  <clippath id="p13d87deb40">
   <rect height="217.44" width="334.8" x="53.328125" y="49.192875"></rect>
  </clippath>
 </defs>
</svg>

</div>

</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We see an equal distribution of positive (<code>True</code>) and negative (<code>False</code>) $1500$-mer training instances (a nice thing to have, but not something we can necessarily count on in the real world!). In fact, the same class proportion and sequence length holds true for the test and validation splits (the former being especially unlikely).</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Machine-Learning">
<a class="anchor" href="#Machine-Learning" aria-hidden="true"><span class="octicon octicon-link"></span></a>Machine Learning<a class="anchor-link" href="#Machine-Learning"> </a>
</h2>
<p>We aim to build a classifer to predict functional motifs given sequences. Formally, we learn a function $f(X)=\vec{y}$ that, given the nature of deep learning networks, fits with non-linearity a probability density function $f=P(y|X;\theta)$. The task is to learn a weights matrix $W\subset \theta$ (a subset since our network may be more complicated) approximating the true class posterior distribution, such that when we call $f(x_{novel})$, we get $\hat{y}\approx y_{true}$, where $y\in \{0,1\} = \{\text{motif,non-motif}\}$.</p>
<h3 id="Task:">
<a class="anchor" href="#Task:" aria-hidden="true"><span class="octicon octicon-link"></span></a>Task:<a class="anchor-link" href="#Task:"> </a>
</h3>
<p>We aim to train a CNN to classify the sequences as positive or negative for homotypic motif clusters. A CNN should be a good choice due to the natural shift-invariance of the architecture. When used on images, CNNs are feature-aware but, importantly, location-agnostic. We want the same approach to motif finding since such subsequences may appear anywhere in the $1500$-mer instances.</p>
<p>One-Hot Basic CNN:</p>

<pre><code>    - CNNs on static one-hot embedded nucleotides.
    - 1 Convolution layer (+ nonlinearity + pooling) and 1 dense (`linear` + activation) output layer.
    - Shift-invariance of the CNN is the main point here.
    - Train the model using stochastic gradient descent and binary cross-entropy loss.

</code></pre>
<p>Embedding with CNN:</p>

<pre><code>    - Use of a learnable embedding on integer-mapped DNA sequences together with the same downstream architecture.

</code></pre>
<p>Transformer:</p>

<pre><code>    - Akin to NLP, we can treat each nucelotide like a token in a sentence.
    - Train an XLNet (with self-attention and all the rest) model with a linear head to predict presence (0/1) of motifs.
    - This particular model is useful since we have somewhat lengthy instances for a transformer ($1500$-mers).</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>I will begin by defining the requisite PyTorch <code>Dataset</code> and <code>Dataloader</code> classes, as well as an extensible <code>LightningModule</code> for use across models.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span> 
</pre></div>

    </div>
</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">from</span> <span class="nn">torch</span> <span class="kn">import</span> <span class="n">nn</span>
<span class="kn">from</span> <span class="nn">torch.nn</span> <span class="kn">import</span> <span class="n">functional</span> <span class="k">as</span> <span class="n">F</span>
<span class="kn">from</span> <span class="nn">torch.utils.data</span> <span class="kn">import</span> <span class="n">Dataset</span><span class="p">,</span> <span class="n">DataLoader</span><span class="p">,</span> <span class="n">random_split</span>
<span class="kn">from</span> <span class="nn">torchvision</span> <span class="kn">import</span> <span class="n">transforms</span>
<span class="kn">import</span> <span class="nn">pytorch_lightning</span> <span class="k">as</span> <span class="nn">pl</span>
<span class="kn">from</span> <span class="nn">pytorch_lightning.metrics.functional</span> <span class="kn">import</span> <span class="n">accuracy</span>
<span class="kn">from</span> <span class="nn">transformers</span> <span class="kn">import</span> <span class="n">AutoConfig</span><span class="p">,</span> <span class="n">AutoModelForSequenceClassification</span> <span class="c1">#the latter includes a linear head.</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The <code>Dataset</code> class is now defined with associated preprocessing methods:</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">class</span> <span class="nc">DNADataset</span><span class="p">(</span><span class="n">Dataset</span><span class="p">):</span>
    <span class="sd">"""DNA / Pytorch dataset."""</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">'one-hot'</span><span class="p">):</span>
        <span class="sd">"""</span>
<span class="sd">        Args:</span>
<span class="sd">            data (np.array): Path to the url .npz file with annotations.</span>
<span class="sd">        """</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">FloatTensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">one_hot_encoding</span><span class="p">(</span><span class="n">X</span><span class="p">))</span> <span class="k">if</span> <span class="n">encoding</span><span class="o">==</span><span class="s1">'one-hot'</span> <span class="k">else</span> <span class="n">torch</span><span class="o">.</span><span class="n">LongTensor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">embedding_encoding</span><span class="p">(</span><span class="n">X</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">y</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">LongTensor</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">transformer_checkpoint</span><span class="o">=</span><span class="n">transformer_checkpoint</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idx</span><span class="p">):</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="p">{</span><span class="s1">'X'</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="p">[</span><span class="n">idx</span><span class="p">],</span> <span class="s1">'y'</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">y</span><span class="p">[</span><span class="n">idx</span><span class="p">]}</span>
        <span class="k">return</span> <span class="n">sample</span>

    <span class="k">def</span> <span class="nf">one_hot_encoding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">seq_array</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
        <span class="sd">"""</span>
<span class="sd">        :param seq_array: np array of DNA sequences</span>
<span class="sd">        :return: np array of one-hot encodings of input DNA sequences of shape N,S,M,</span>
<span class="sd">                where N is the number of instances, S is the cardinality of the corpus,</span>
<span class="sd">                and M is the length of each sequence.</span>
<span class="sd">        """</span>
        <span class="n">nuc2id</span> <span class="o">=</span> <span class="p">{</span><span class="s1">'A'</span> <span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">'C'</span> <span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">'T'</span> <span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">'G'</span> <span class="p">:</span> <span class="mi">3</span><span class="p">}</span>

        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_array</span><span class="p">)</span>
        <span class="n">S</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">nuc2id</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_array</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c1">#since we know each sequence is the same length</span>

        <span class="n">onehot_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">N</span><span class="p">,</span> <span class="n">S</span><span class="p">,</span> <span class="n">M</span><span class="p">))</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s1">'Nucelotides -&gt; One-Hot'</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">seq_num</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">seq_array</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">seq_idx</span><span class="p">,</span> <span class="n">nucleotide</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
                <span class="n">nuc_idx</span> <span class="o">=</span> <span class="n">nuc2id</span><span class="p">[</span><span class="n">nucleotide</span><span class="p">]</span>
                <span class="n">onehot_array</span><span class="p">[</span><span class="n">seq_num</span><span class="p">,</span> <span class="n">nuc_idx</span><span class="p">,</span> <span class="n">seq_idx</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">onehot_array</span>

    <span class="k">def</span> <span class="nf">embedding_encoding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_array</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
        <span class="sd">"""</span>
<span class="sd">        :param seq_array: np array of DNA sequences</span>
<span class="sd">        :return: np array of integer encodings of input DNA sequences</span>
<span class="sd">        """</span>
        <span class="n">nuc2id</span> <span class="o">=</span> <span class="p">{</span><span class="s1">'A'</span> <span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">'C'</span> <span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">'T'</span> <span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">'G'</span> <span class="p">:</span> <span class="mi">3</span><span class="p">}</span>
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_array</span><span class="p">)</span>
        <span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_array</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c1">#since we know each sequence is the same length</span>

        <span class="n">int_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">((</span><span class="n">N</span><span class="p">,</span><span class="n">M</span><span class="p">),</span><span class="mi">0</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">'Nucleotides -&gt; Integers'</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">seq_num</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">seq_array</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">seq_idx</span><span class="p">,</span> <span class="n">nucleotide</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
                <span class="n">nuc_idx</span> <span class="o">=</span> <span class="n">nuc2id</span><span class="p">[</span><span class="n">nucleotide</span><span class="p">]</span>
                <span class="n">int_array</span><span class="p">[</span><span class="n">seq_num</span><span class="p">,</span> <span class="n">seq_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">nuc2id</span><span class="p">[</span><span class="n">nucleotide</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">int_array</span>

<span class="k">class</span> <span class="nc">DNADataModule</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">LightningDataModule</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">64</span><span class="p">,</span><span class="n">encoding</span><span class="o">=</span><span class="s1">'one-hot'</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">=</span> <span class="n">batch_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_fpath</span> <span class="o">=</span> <span class="n">data_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">encoding</span><span class="o">=</span><span class="n">encoding</span>

    <span class="k">def</span> <span class="nf">setup</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stage</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">all_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data_fpath</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">train</span> <span class="o">=</span> <span class="n">DNADataset</span><span class="p">(</span><span class="n">all_data</span><span class="p">[</span><span class="s1">'train_seq'</span><span class="p">],</span><span class="n">all_data</span><span class="p">[</span><span class="s1">'train_y'</span><span class="p">],</span><span class="bp">self</span><span class="o">.</span><span class="n">encoding</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">val</span>   <span class="o">=</span> <span class="n">DNADataset</span><span class="p">(</span><span class="n">all_data</span><span class="p">[</span><span class="s1">'train_seq'</span><span class="p">],</span><span class="n">all_data</span><span class="p">[</span><span class="s1">'train_y'</span><span class="p">],</span><span class="bp">self</span><span class="o">.</span><span class="n">encoding</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">test</span>  <span class="o">=</span> <span class="n">DNADataset</span><span class="p">(</span><span class="n">all_data</span><span class="p">[</span><span class="s1">'test_seq'</span><span class="p">],</span> <span class="n">all_data</span><span class="p">[</span><span class="s1">'test_y'</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">encoding</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">train_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">DataLoader</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">train</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">val_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">DataLoader</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">val</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">test_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">DataLoader</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">test</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">dm_onehot</span> <span class="o">=</span> <span class="n">DNADataModule</span><span class="p">(</span><span class="n">data_dir</span><span class="o">=</span><span class="s1">'./data_dna_cnn/hw1_data.npz'</span><span class="p">,</span><span class="n">encoding</span><span class="o">=</span><span class="s1">'one-hot'</span><span class="p">)</span>
<span class="n">dm_onehot</span><span class="o">.</span><span class="n">setup</span><span class="p">()</span>
<span class="n">dm_int</span> <span class="o">=</span> <span class="n">DNADataModule</span><span class="p">(</span><span class="n">data_dir</span><span class="o">=</span><span class="s1">'./data_dna_cnn/hw1_data.npz'</span><span class="p">,</span><span class="n">encoding</span><span class="o">=</span><span class="s1">'int'</span><span class="p">)</span> <span class="c1">#for use with 2nd CNN and Transformer</span>
<span class="n">dm_int</span><span class="o">.</span><span class="n">setup</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Nucelotides -&gt; One-Hot
Nucelotides -&gt; One-Hot
Nucelotides -&gt; One-Hot
Nucleotides -&gt; Integers
Nucleotides -&gt; Integers
Nucleotides -&gt; Integers
</pre>
</div>
</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Note that for the Transformer, I will be pulling an untrained model from Huggingface. That involves invoking the <code>config</code> of the desired model, but instantiating the model itself from the config only (<strong><em>not</em></strong> <code>from_pretrained</code>).</p>
<p>As for the tokenizer, we don't really need the actual model's tokenizer with associated vocabulary, etc., since we can just map each nucleotide to an integer. This should be equivalent to character-level tokenization of each nucleotide with an untrained vocabulary, but also allows for the use of <code>nn.Embedding()</code> for the second CNN.</p>
<p>For illustrative purposes:</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">AutoModelForSequenceClassification</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span>
    <span class="n">AutoConfig</span><span class="o">.</span><span class="n">from_pretrained</span><span class="p">(</span><span class="s1">'xlnet-base-cased'</span><span class="p">)</span>
    <span class="p">)</span>
<span class="n">model</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>XLNetForSequenceClassification(
  (transformer): XLNetModel(
    (word_embedding): Embedding(32000, 768)
    (layer): ModuleList(
      (0): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (1): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (2): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (3): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (4): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (5): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (6): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (7): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (8): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (9): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (10): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (11): XLNetLayer(
        (rel_attn): XLNetRelativeAttention(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (ff): XLNetFeedForward(
          (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (layer_1): Linear(in_features=768, out_features=3072, bias=True)
          (layer_2): Linear(in_features=3072, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
    )
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (sequence_summary): SequenceSummary(
    (summary): Linear(in_features=768, out_features=768, bias=True)
    (first_dropout): Identity()
    (last_dropout): Dropout(p=0.1, inplace=False)
  )
  (logits_proj): Linear(in_features=768, out_features=2, bias=True)
)</pre>
</div>

</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The transformer model is currently instantiated for classification (<code>model.logits_proj</code> has <code>out_features=2</code>). Note that we will need to take some form of activation across these logits (sigmoid, or even plain argmax). We can use <code>BCEWithLogits</code> (binary cross entropy with logits) to accomplish this.</p>
<p>Note due to the nature of binary (versus multlabel) classification, that we could instead have a <em>single</em> neuron output and use sigmoidal activation (analagous to logistic regression).</p>
<p>The key difference here is that a single neuronal output will evaluate the probability of a single output together, rather than the probability of each class independently. By keeping the output dimensions as $2$, we gain a small amount out of the network (each output gets its own weight) that would be implicit otherwise.</p>
<p>However, the dimensions of the label space need to pairwise match. Since labels are currently set up in the dataset with cardinality of 1, I will instead just refactor the model slightly to have a single output dimension. It's not terribly clear what the advantage in this toy example would be to changing or keeping the architecture, at any rate.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">logits_proj</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">Linear</span><span class="p">(</span><span class="n">in_features</span><span class="o">=</span><span class="mi">768</span><span class="p">,</span><span class="n">out_features</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">bias</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">logits_proj</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>Linear(in_features=768, out_features=1, bias=True)</pre>
</div>

</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered">
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Perfect! This will be incorporated below.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">class</span> <span class="nc">DNACNNEmbedding</span><span class="p">(</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">DNACNNEmbedding</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">conv1</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">Conv2d</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">conv2</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">Conv2d</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">F</span><span class="o">.</span><span class="n">relu</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">conv1</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">F</span><span class="o">.</span><span class="n">relu</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">conv2</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>

<span class="k">class</span> <span class="nc">DNACNNEmbedding</span><span class="p">(</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">Model</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">conv1</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">Conv2d</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">conv2</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">Conv2d</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">F</span><span class="o">.</span><span class="n">relu</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">conv1</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">F</span><span class="o">.</span><span class="n">relu</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">conv2</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>

<span class="k">class</span> <span class="nc">DNATransformer</span><span class="p">(</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">transformer_checkpoint</span><span class="o">=</span><span class="s1">'xlnet-base-cased'</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">DNATransformer</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        
        <span class="c1">#Load untrained model using checkpointed config.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">=</span><span class="n">AutoConfig</span><span class="o">.</span><span class="n">from_pretrained</span><span class="p">(</span><span class="n">transformer_checkpoint</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">AutoModelForSequenceClassification</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">logits_proj</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">Linear</span><span class="p">(</span><span class="n">in_features</span><span class="o">=</span><span class="mi">768</span><span class="p">,</span><span class="n">out_features</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">bias</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">x</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="k">class</span> <span class="nc">DNALitModule</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">LightningModule</span><span class="p">):</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">net</span><span class="p">:</span><span class="n">pl</span><span class="o">.</span><span class="n">LightningModule</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">net</span> <span class="o">=</span> <span class="n">net</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">criterion</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BCEWithLogitsLoss</span><span class="p">()</span> <span class="c1">#Sigmoid--&gt;BCE Loss.</span>

    <span class="k">def</span> <span class="nf">training_step</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch</span><span class="p">,</span> <span class="n">batch_idx</span><span class="p">):</span>
        <span class="c1"># --------------------------</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">batch</span>
        <span class="n">y_hat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">net</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="c1">#take the most probable index.</span>
        <span class="n">loss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">criterion</span><span class="p">(</span><span class="n">y_hat</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="s1">'train_loss'</span><span class="p">,</span> <span class="n">loss</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">loss</span>
        <span class="c1"># --------------------------</span>

    <span class="k">def</span> <span class="nf">validation_step</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch</span><span class="p">,</span> <span class="n">batch_idx</span><span class="p">):</span>
        <span class="c1"># --------------------------</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">batch</span>
        <span class="n">y_hat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">net</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">loss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">criterion</span><span class="p">(</span><span class="n">y_hat</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="s1">'val_loss'</span><span class="p">,</span> <span class="n">loss</span><span class="p">)</span>
        <span class="c1"># --------------------------</span>

    <span class="k">def</span> <span class="nf">test_step</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch</span><span class="p">,</span> <span class="n">batch_idx</span><span class="p">):</span>
        <span class="c1"># --------------------------</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">batch</span>
        <span class="n">y_hat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">net</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">loss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">criterion</span><span class="p">(</span><span class="n">y_hat</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="s1">'test_loss'</span><span class="p">,</span> <span class="n">loss</span><span class="p">)</span>
        <span class="c1"># --------------------------</span>

    <span class="k">def</span> <span class="nf">configure_optimizers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">optimizer</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">optim</span><span class="o">.</span><span class="n">Adam</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="p">(),</span> <span class="n">lr</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">optimizer</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

</div>



  </div><!-- from https://github.com/utterance/utterances -->
<script src="https://utteranc.es/client.js"
        repo="simonlevine/simonsays"
        issue-term="title"
        label="blogpost-comment"
        theme="github-light"
        crossorigin="anonymous"
        async>
</script><a class="u-url" href="/simonsays/jupyter/2021/02/17/DNA_CNN.html" hidden></a>
</article>
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/simonsays/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">
        <p class="feed-subscribe">
          <a href="/simonsays/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/simonsays/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p>
      </div>
      <div class="footer-col">
        <p>Tutorials and Musings.</p>
      </div>
    </div>

    <div class="social-links"><ul class="social-media-list"><li><a rel="me" href="https://github.com/simonlevine" title="simonlevine"><svg class="svg-icon grey"><use xlink:href="/simonsays/assets/minima-social-icons.svg#github"></use></svg></a></li></ul>
</div>

  </div>

</footer>
</body>

</html>
